### ISSUE TITLE
Maze solve

### ISSUE THREAD

**Issue by AntoraDev** [AntoraDev/Cifar100#1]
% Define the maze layout
maze = [1, 1, 1, 1, 1; 
        1, -1, 1, -1, 1; 
        1, 1, 1, 1, 1; 
        1, -1, 1, -1, 1; 
        1, 1, 1, 1, 1];

% Define the agent's starting position
startPosition = [1, 3];

% Define the goal position
goalPosition = [3, 1];

% Define the available actions (up, down, left, right)
actions = [0 -1; -1 0; 0 1; 1 0];

% Set hyperparameters
alpha = 0.1; % Learning rate
gamma = 0.95; % Discount factor
epsilon = 1; % Begins with 1 having a lot of explorations
episodes = 100; % How many episodes for training

% Initialize the Q-table and the cumulativeReward array
cumulativeReward = zeros(1, episodes);

% Initialize array to store Q-values for the start position and a chosen action (e.g., 'up')
qValueTrack = zeros(episodes, 1);

% Initialize the Q-table
numStates = numel(maze);
numActions = size(actions, 1);
Q = rand(numStates, numActions) * 0.01;

% Define rewards
goalReward = 100;
obstacleReward = -2;
stepReward = -1;

% Define the reward matrix
rewardMatrix = ones(size(maze)) * stepReward;
rewardMatrix(maze == -1) = obstacleReward;
rewardMatrix(goalPosition(1), goalPosition(2)) = goalReward;

% Convert maze position to state index
stateIndex = @(position) (position(1) - 1) * size(maze, 2) + position(2);

% Create the figure and axes for the maze
figure;
axes('XLim',[1 size(maze, 1)], 'YLim',[1 size(maze, 2)]);
hold on;

% Plot the maze
for row = 1:size(maze, 1)
    for col = 1:size(maze, 2)
        if maze(row, col) == -1
            rectangle('Position', [col-0.5, row-0.5, 1, 1], 'FaceColor', 'white');
        elseif maze(row, col) == 1
            rectangle('Position', [col-0.5, row-0.5, 1, 1], 'FaceColor', 'black');
        end
    end
end

% Plot the goal position
rectangle('Position', [goalPosition(2)-0.5, goalPosition(1)-0.5, 1, 1], 'FaceColor', 'white');

% Plot the agent's starting position
agentHandle = rectangle('Position', [startPosition(2)-0.5, startPosition(1)-0.5, 1, 1], 'Curvature', [1, 1], 'FaceColor', 'red');

% Start the Q-learning loop
for episode = 1:episodes
    % Initialize the agent's position and the episode's reward
    currentPosition = startPosition;

    % Iterate within each episode until reaching the goal
    while ~isequal(currentPosition, goalPosition)

        % Convert the current position to state index
        currentState = stateIndex(currentPosition);

        % Choose an action using epsilon-greedy strategy
        if rand < epsilon
            action = randi(numActions);
        else
            [~, action] = max(Q(currentState, :));
        end

        % Get the new position based on the chosen action
        newPosition = currentPosition + actions(action, :);

        % Check if the new position is valid
        if newPosition(1) < 1 || newPosition(1) > size(maze, 1) || newPosition(2) < 1 || newPosition(2) > size(maze, 2) || maze(newPosition(1), newPosition(2)) == -1
            newPosition = currentPosition;
        end

        % Convert the new position to state index
        newState = stateIndex(newPosition);

        % Get the reward for the new position
        reward = rewardMatrix(newPosition(1), newPosition(2));

        % Update the Q-value for the current state and action
        Q(currentState, action) = Q(currentState, action) + ...
                                  alpha * (reward + gamma * max(Q(newState, :)) - Q(currentState, action));

        % Update the cumulative reward
        cumulativeReward(episode) = cumulativeReward(episode) + reward;

        % Update the agent's position
        currentPosition = newPosition;

        % Update the agent's graphical representation
        set(agentHandle, 'Position', [currentPosition(2)-0.5, currentPosition(1)-0.5, 1, 1]);
        drawnow;

        % Pause for a short time to allow for visualization of the agent's movement
        %pause(0.05);
    end

        % Record the Q-value for the start position and 'up' action at the end of the episode
        qValueTrack(episode) = Q(stateIndex(startPosition), 1);  % 1 is the index for 'up' action

    % Decrease epsilon for the next episode
    epsilon = max(0.1, epsilon * 0.99);
end

% Plotting the cumulative reward to understand the agent's learning
figure;
plot(1:episodes, cumulativeReward);
title('Cumulative Reward vs Episodes');
xlabel('Episodes');
ylabel('Cumulative Reward');

% The agent has completed training
disp('Training completed!');

% Display the learned Q-values
disp('Learned Q-values:');
disp(Q);

% Plotting the tracked Q-value to understand the Q-value evolution
figure;
plot(1:episodes, qValueTrack);
title('Q-value Evolution');
xlabel('Episodes');
ylabel('Q-value');

maze1 = [ 1  1  1   1  1;
         1  -1  1  -1  1;
         1   1  1  1  1;
         1  -1  1  -1  1;
         1  1  1  1  1];

% Test run for 3 times
for test_run = 1:3
   
    % Assign the starting position for each test run
    startPosition = getRandomStart(maze1);
   
    % Print the Starting and Goal Position
    fprintf('\nTest Run %d\n', test_run);
    fprintf('Starting Position: [%d, %d]\n', startPosition(1), startPosition(2));
    fprintf('Goal Position: [%d, %d]\n', goalPosition(1), goalPosition(2));
   
    % Initialize the agent's current position
    currentPosition = startPosition;
   
    % Iterate within each test run until reaching the goal
    while ~isequal(currentPosition, goalPosition)
       
        % Print the Current Position
        fprintf('Current Position: [%d, %d]\n', currentPosition(1), currentPosition(2));
       
        % Convert the current position to state index
        currentState = stateIndex(currentPosition);
       
        % Choose the action with the highest Q-value for the current state
        [~, actionIndex] = max(Q(currentState, :));
% Choose the best action based on the highest Q-value
[~, actionIndex] = max(Q(currentState, :));

% Compute the new (next) position based on the chosen action
newPosition = currentPosition + actions(actionIndex, :);

    % Check if the new (next) position is a valid position in the maze
    isValid = newPosition(1) >= 1 && newPosition(1) <= size(maze1, 1) && newPosition(2) >= 1 && newPosition(2) <= size(maze1, 2);
   
    % Check if the new (next) position is valid (within the maze boundaries and not an obstacle)
    if isValid && maze1(newPosition(1), newPosition(2)) ~= -1
       
        % Update the agent's position (Current position now becomes the valid Next position)
        currentPosition = newPosition;
    else
        fprintf('Invalid Next Position, retrying...\n');
        % The new (next) position is invalid or an obstacle;
        % The agent stays in the current position and re-evaluates action in the next loop iteration
        continue;
    end
    end
   
    % The test run is completed
    fprintf('Reached Goal Position: [%d, %d]!\n', currentPosition(1), currentPosition(2));
    disp('Test run completed!\n');
end
